### DISCUSSION TITLE
Nova Actions, i`m frustrated - empty resource on click

### DISCUSSION THREAD

**Discussion by jansgescheit** [laravel/nova-issues#5647] [Q&A] [answered]
Sorry I'm a little frustrated with Nova.

I define an action on a Nova resource and the `public function actions(NovaRequest $request): array` method. 
Now I have the following problems where I doubt myself a little bit and I can`t find anything in the documentation.

In the index view I can access `$this` / `$this->resource` in the said method and get a filled eloquent model for the respective row of the the resource.  I have the same behavior in the detail view. However, as soon as I click on an action in the detail view, the action method of the Nova Resource is executed again and `$this` contains an eloquent model, but this is an empty one.  At this point the context is somehow forgotten and I have to restore it manually.

This confuses me and I am now reading the model data separately to use it:

```php
class Post extends Resource
{
    public static $model = \App\Models\Post::class;

    [...]

    public function actions(NovaRequest $request): array
    {
        /** @var Post $resource */
        $resource = $request->selectedResources()->first() ?? $this; 

        // index $this is filled
        // detail $this is filled
        // detail + Action Click $this is not filled

        return [
            EMailAction::make($this->locale), // wont work on Detail + Action Click,
            EMailAction::make($resource->locale), // will work on Detail + Action Click, but not on index if more than one is selected
        ];
    }
}
```

As I said this is very confusing to think in different dimensions (Index/Detail/Click/Index Select) here. Why doesn't Nova fill the `$this` object correctly here? 

I believe that there are more side effects that also do not work as expected. Is this now a bug a feature?  Am i doing something wrong?

**Comment by crynobone** [comment, marked as answer]
> EMailAction::make($resource->locale), // will work on Detail + Action Click, but not on index if more than one is selected

Obviously, this would cause an issue when `locale` can now be multiple (different for each resource) but there's only a single instance of `EMailAction`. You should only allow this action to be selected for following usage:

1. Single resource selected from Index
2. Inline action button
3. Detail page

```php
EMailAction::make()->canSee(function (NovaRequest $request) {
    return $request->selectedResourceIds()?->count() === 1;
})->showInline()->showOnDetail(),
```

```php
    /**
     * Perform the action on the given models.
     *
     * @param  \Laravel\Nova\Fields\ActionFields  $fields
     * @param  \Illuminate\Support\Collection  $models
     * @return mixed
     */
    public function handle(ActionFields $fields, Collection $models)
    {
        $locale = $models->first()->locale;
```

**Reply by crynobone** [reply]
If you need to access the value from `fields()`, you should be able to get the current resource using `$request->selectedResources()->first()` or `$request->findModel()`

**Reply by jansgescheit** [reply]
thanks for your tipps!

Inside of the `field()` method `$request->selectedResources()->first()` I get the expected data set. With the method `$request->findModel()` I get only a new eloquent model without data. 

This I think is for someone who is not 100% into Nova, just has the documentation. Where I think it still lacks some information just how the lifecycle looks like, very difficult to think of all possible dimensions here.

It would be nice from the expectation if this was somehow uniformly designed. Especially since the `action()` as well as `fields()` method is currently already called for each individual model.

**Reply by crynobone** [reply]
> It would be nice from the expectation if this was somehow uniformly designed. Especially since the action() as well as fields() method is currently already called for each individual model.

Actions were designed to handle multiple resources first and that design persist until now. As explained earlier it always recommended to access `$model` from `handle()` method where you can safely handle multiple models or a single model without any issue.

**Reply by crynobone** [reply]
Trying to access multiple resources and handle everything under `fields()` is where you get all these issues. So limit selected based on the above suggestion.

**Comment by crynobone** [comment]
New in 4.26.0 you can simplify the above suggestion using `sole()` 

```php
EMailAction::make()->sole(),
```
